src/library_pthread.js

```diff
@@ -121,6 +121,91 @@ var LibraryPThread = {
 #endif
     },
 
+    initMessageRelay: function() {
+      // Spawn a dedicated worker for passing messages between threads. Instead
+      // of having each thread hold a message port for every other thread, they
+      // can just send messages with a `targetThread` property to the relay and
+      // the relay will forward the message to the intended recipient.
+      // Alternatively, the main thread could play this role, but then messages
+      // could be held up while the main thread is busy with other work.
+      var relayCode = '';
+
+      // On Node, first do the minimal work to get a Web-compatible messaging
+      // interface. Use a template literal to avoid JS parsers failing on the
+      // unexpected `import` token.
+#if ENVIRONMENT_MAY_BE_NODE
+      if (ENVIRONMENT_IS_NODE) {
+        relayCode +=
+`(await import('node:worker_threads'))
+  .parentPort
+  .on('message', (data) => onmessage({ data: data }));
+Object.assign(global, {
+  self: global,
+});
+`;
+      }
+#endif
+
+      relayCode += '(' + (() => {
+        // Map pthread IDs to message ports we use to communicate with those
+        // pthreads.
+        const threadPorts = new Map();
+
+        // Map recipient pthread IDs for whom we don't yet have a message port
+        // to messages we've received for them.
+        const bufferedMessages = new Map();
+
+        function handleMessage(msg) {
+          const thread = msg.data.targetThread;
+          const port = threadPorts.get(thread);
+          if (port !== undefined) {
+            port.postMessage(msg.data, msg.data.transferList);
+          } else {
+            // Hold on to the message until we receive a port for the recipient.
+            if (!bufferedMessages.has(thread)) {
+              bufferedMessages.set(thread, []);
+            }
+            bufferedMessages.get(thread).push(msg);
+          }
+        }
+
+        self.onmessage = (msg) => {
+          const cmd = msg.data.cmd;
+          const thread = msg.data.thread;
+          if (cmd === 'create') {
+            const port = msg.data.port;
+            threadPorts.set(thread, port);
+            port.onmessage = handleMessage;
+            // Forward any messages we have already received for this thread.
+            if (bufferedMessages.has(thread)) {
+              bufferedMessages.get(thread).forEach(handleMessage);
+              bufferedMessages.delete(thread);
+            }
+            return;
+          }
```

Comment:
`else if` instead of return here?  .. and then maybe assert in the final `else`?

Reply:
We don't have `assert` available in this worker, but I will print an error at least.